MetaDataset.py

```diff
@@ -885,23 +884,140 @@ def init_seq_order(self, epoch=None, seq_list=None):
+    full_epoch = (epoch - 1) // self.partition_epoch + 1 if epoch else 1
+    self.rnd.seed(full_epoch)
 
     if not need_reinit:
+      self._num_seqs = num_seqs_saved
       return False
 
-    # First init sequence order for sub-datasets as usual to get a list of available sequences. This way sorting and
+    # First init sequence order for sub-datasets as usual to get a list of available sequences. This way, sorting and
     # partition epoch of the individual sub-datasets is still supported. Later we will call init_seq_order again with a
     # sequence list to e.g. apply joint sorting or partition epoch of all sequences.
     for dataset in self.datasets.values():
       dataset.init_seq_order(epoch=epoch)
 
-    self.dataset_sorted_seq_idx_list = []  # We will fill this as we go
-    self.used_num_seqs_per_subset = [0] * len(self.datasets)
+    try:
+      total_num_seqs = sum([self.datasets[k].num_seqs for k in sorted(self.datasets.keys())])
+    except Exception:
+      total_num_seqs = None
+
+    if total_num_seqs is not None:
+      self.used_num_seqs_per_subset = [
+        self.datasets[self.dataset_idx2key_map[dataset_idx]].num_seqs for dataset_idx in range(len(self.datasets))]
+      self.dataset_seq_idx_boundaries = self._create_dataset_seq_idx_boundaries()
+
+      if self.seq_ordering == "random_dataset":
+        seq_order = self._get_random_dataset_seq_order(total_num_seqs)
+      else:
+        seq_order = self.get_seq_order_for_epoch(
+          epoch=epoch, num_seqs=total_num_seqs, get_seq_len=self._get_seq_length)
+      self._num_seqs = len(seq_order)
+
+      # Apply seq_order to self.dataset_seq_idx.
+      # We have to re-calculate the seq_idx's because we will sort the datasets below.
+      self.dataset_sorted_seq_idx_list = []
+      counters = [0] * len(self.datasets)
+      for seq_idx in seq_order:
+        dataset_idx, _ = self._seq_idx_to_dataset_seq_idx(seq_idx)
+        dataset_seq_idx = counters[dataset_idx]
+        counters[dataset_idx] += 1
+        self.dataset_sorted_seq_idx_list.append((dataset_idx, dataset_seq_idx))
+
+      # We only want to load those sequences in the sub-datasets that appear in seq_order. For this, we extract
+      # sequence lists containing the subset of sequences for each dataset from seq_order.
+      seq_lists = [[] for _ in self.datasets]
+      for seq_idx in seq_order:
+        dataset_idx, dataset_seq_idx = self._seq_idx_to_dataset_seq_idx(seq_idx)
+        dataset = self.datasets[self.dataset_idx2key_map[dataset_idx]]
+        seq_tag = dataset.get_tag(dataset_seq_idx)
+        seq_lists[dataset_idx].append(seq_tag)
+
+      # It may be large, so better delete it early, we don't need it anymore.
+      del seq_order
+
+      # Re-initialize sequence orders of sub-datasets with created sequence list.
+      for dataset_idx, dataset_key in self.dataset_idx2key_map.items():
+        self.datasets[dataset_key].init_seq_order(epoch=epoch, seq_list=seq_lists[dataset_idx])
```

Comment:
Can it happen that you init some datasets twice / multiple times here? (If the `dataset_key` is used multiple times.)

Reply:
Ah sorry, no, this should be unique.